storage/innobase/trx/trx0sys.cc

```diff
@@ -470,7 +470,24 @@ void trx_sys_update_wsrep_checkpoint(
         the same xid_seqno.
         Allow current and new values be the same, without introducing new flags
         and logic to prevent double storing of the same value */
-        ut_ad(xid_seqno >= trx_sys_cur_xid_seqno);
+
+        /* Case 3:
+
+           NBO allows interleaving of non-conflicting transactions. This allows
+           other transactions to be committed in InnoDB while there is an
+           ongoing NBO update. This is possible when the NBO thread has
+           finished phase one but is yet to start its phase two. Before adding
+           the below NBO condition, the other transactions updated its XID
+           there by resulting in the NBO thread to hit the below assertion as
+           it finds that the current thread no more has the latest xid in
+           InnoDB.
+
+           i.e,
+           Seqno X - NBO begin
```

Comment:
I'm confused about this explanation. If NBO end tries to store X+N+1, there should not be any assertion.

Reply:
Here, we get assertion when NBO tries to store X, the one we got during phase one. I have changed the comment to avoid confusion.